lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,342 @@
+    }
+    newArgAttrs.push_back(DictionaryAttr::get(op->getContext(), newArgAttr));
+  }
+
+  return ArrayAttr::get(op->getContext(), newArgAttrs);
+};
+
+void GrandCentralPass::runOnOperation() {
+  CircuitOp circuitOp = getOperation();
+
+  AnnotationSet annotations(circuitOp);
+  if (annotations.empty())
+    return;
+
+  auto context = &getContext();
+  OpBuilder builder(context);
+  builder.setInsertionPointToEnd(getOperation()->getBlock());
+
+  // Store a mapping of interface name to InterfaceOp.
+  llvm::StringMap<sv::InterfaceOp> interfaces;
+
+  // Track the order that interfaces should be emitted in.
+  llvm::SmallVector<std::pair<StringRef, StringRef>> interfaceKeys;
+
+  // Examine the Circuit's Annotations doing work to remove Grand Central
+  // Annotations.  Ignore any unprocesssed annotations and rewrite the Circuit's
+  // Annotations with these when done.
+  llvm::SmallVector<Attribute> unprocessedAnnos;
+  for (auto anno : annotations) {
+    auto clazz = anno.getClass();
+
+    if (clazz == "sifive.enterprise.grandcentral.AugmentedBundleType") {
+      auto defName = anno.getMember<StringAttr>("defName");
+      for (auto elt : anno.getMember<ArrayAttr>("elements")) {
+        auto eltDict = elt.cast<DictionaryAttr>();
+        auto name = eltDict.getAs<StringAttr>("name").getValue();
+        auto tpe = eltDict.getAs<StringAttr>("tpe").getValue();
+        StringRef description = {};
+        if (auto maybeDescription = eltDict.getAs<StringAttr>("description"))
+          description = maybeDescription.getValue();
+
+        auto key = std::make_pair(defName.getValue(), name);
+        if (tpe == "sifive.enterprise.grandcentral.AugmentedBundleType")
+          interfaceMap[key] = {Component::Type::Bundle, description};
+        else if (tpe == "sifive.enterprise.grandcentral.AugmentedVectorType")
+          interfaceMap[key] = {Component::Type::Vector, description};
+        else
+          interfaceMap[key] = {Component::Type::Ground, description};
+        interfaceKeys.push_back(key);
+      }
+      // Create the interfaces.  These will be populated later.
+      if (interfaces.count(defName.getValue()) == 0) {
+        auto interfaceOp =
+            builder.create<sv::InterfaceOp>(circuitOp->getLoc(), defName);
+        ImplicitLocOpBuilder blockBuilder(interfaceOp->getLoc(),
+                                          interfaceOp.getRegion());
+        blockBuilder.createBlock(&interfaceOp.getRegion());
+        blockBuilder.create<sv::TypeDeclTerminatorOp>();
+        interfaces[defName.getValue()] = interfaceOp;
+      }
+      continue;
+    }
+
+    unprocessedAnnos.push_back(anno.getDict());
+  }
+
+  circuitOp->setAttr("annotations", builder.getArrayAttr(unprocessedAnnos));
+
+  // Walk through the circuit to collect additional information.  If this fails,
+  // signal pass failure.
+  for (auto &op : circuitOp.getBody()->getOperations()) {
+    bool failed =
```

Comment:
This one is tricky as I need the type of the `op` to be casted to `FModuleOp` or `FExtModuleOp` so that the correct visitor method gets called. (This is effectively what the expansion of `TypeSwitch` is doing anyway.) I could change this into two casts and checks, but it's probably fine as is.

Reply:
I can refactor this to clean it up. I'm kind of doing the wrong thing here with overloading `visitDecl` to handle modules. Working on this...